internal/tools/semconvkit/main.go

```diff
@@ -58,16 +130,156 @@ func render(src, dest string, data *SemanticConventions) error {
+		info, err = os.Stat(sub)
+		if err == nil && info.IsDir() {
+			return hint, nil
+		}
 	}
 
-	sc := &SemanticConventions{TagVer: *tag}
+	v, err := semver.NewVersion(curr)
+	if err != nil {
+		return "", fmt.Errorf("invalid current version %q: %w", curr, err)
+	}
 
-	if err := render("templates/*.tmpl", *out, sc); err != nil {
-		log.Fatal(err)
+	entries, err := os.ReadDir(root)
+	if err != nil {
+		return "", fmt.Errorf("error reading root %q: %w", root, err)
+	}
+
+	var prev *semver.Version
+	for _, entry := range entries {
+		slog.Debug("root entry", "name", entry.Name())
+		if !entry.IsDir() {
+			continue
+		}
+
+		ver, err := semver.NewVersion(entry.Name())
+		if err != nil {
+			slog.Debug("not a version dir", "name", entry.Name())
+			// Ignore errors for non-semver directories.
+			continue
+		}
+		slog.Debug("found version dir", "prev", ver)
+		if ver.LessThan(v) && (prev == nil || ver.GreaterThan(prev)) {
+			slog.Debug("new previous version", "version", ver)
+			prev = ver
+		}
+	}
+
+	if prev == nil {
+		return "", errors.New("no previous version found")
+	}
+	return prev.Original(), nil
+}
+
+// SemanticConventions are information about the semantic conventions being
+// generated.
+type SemanticConventions struct {
+	// TagVer is the tagged version (i.e. v.7.0 and not 1.7.0).
+	TagVer string
+}
+
+func (sc SemanticConventions) SemVer() string {
+	return strings.TrimPrefix(*tag, "v")
+}
+
+// Migration contains the details about the migration from the previous
+// semantic conventions to the current one.
+type Migration struct {
+	CurrVer  string
+	PrevVer  string
+	Removals []string
+	Renames  []Rename
+}
+
+// Remove is a semantic convention declaration that has been renamed.
+type Rename struct {
+	Old, New string
+}
+
+func newMigration(curr, prev string) (*Migration, error) {
+	cDecl, err := decls.GetNames(curr, parse)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing current version %q: %w", curr, err)
+	}
+
+	pDecl, err := decls.GetNames(prev, parse)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing previous version %q: %w", prev, err)
+	}
+
+	m := Migration{
+		CurrVer:  filepath.Base(curr),
+		PrevVer:  filepath.Base(prev),
+		Removals: inAnotB(pDecl, cDecl),
+		Renames:  renames(pDecl, cDecl),
+	}
+
+	sort.Strings(m.Removals)
+	sort.Slice(m.Renames, func(i, j int) bool {
+		return m.Renames[i].Old < m.Renames[j].Old
+	})
+
+	return &m, nil
+}
+
+func parse(d ast.Decl) []string {
+	var out []string
+	switch decl := d.(type) {
+	case *ast.FuncDecl:
+		out = []string{decl.Name.Name}
+	case *ast.GenDecl:
+		if decl.Tok == token.CONST || decl.Tok == token.VAR {
+			for _, spec := range decl.Specs {
+				if valueSpec, ok := spec.(*ast.ValueSpec); ok {
+					for _, name := range valueSpec.Names {
+						out = append(out, name.Name)
+					}
+				}
+			}
+		}
+	}
+	return out
+}
+
+// inAnotB returns the canonical names in a that are not in b.
+func inAnotB(a, b decls.Names) []string {
```

Comment:
maybe rename to `diff` as this is a set difference

Reply:
> a exclusive of b, which is not a set difference

I think it is but I have not studied Maths in English. Therefore, I trust you more than [Wikipedia](https://en.m.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement).

Feel free to resolve. I think the current name is more straightforward and obvious.